File: packages/miner/src/hash.js
import { keccak_256 } from "@noble/hashes/sha3";

export const MAX_U256 = (1n << 256n) - 1n;

/** Byte length of abi.encodePacked(bytes32, address, uint256). */
export const PREIMAGE_BYTES = 32 + 20 + 32;

const NONCE_OFFSET = 52;

function stripHex(hex) {
  return hex.startsWith("0x") || hex.startsWith("0X") ? hex.slice(2) : hex;
}

/** Parse a hex string into exactly `length` bytes, left-padding with zeros. */
export function hexToBytes(hex, length) {
  let clean = stripHex(hex);
  if (clean.length % 2 !== 0) clean = "0" + clean;
  if (clean.length > length * 2) {
    throw new Error(`expected at most ${length} bytes, got ${clean.length / 2}`);
  }
  const out = new Uint8Array(length);
  const offset = length - clean.length / 2;
  for (let i = 0; i < clean.length / 2; i++) {
    out[offset + i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

export function bytesToHex(bytes) {
  let s = "0x";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

export function bytesToBigInt(bytes) {
  let n = 0n;
  for (const b of bytes) n = (n << 8n) | BigInt(b);
  return n;
}

/**
 * Build the reusable 84-byte preimage buffer for one (challenge, miner) pair.
 * The nonce is written in place per attempt so the hot loop allocates nothing.
 */
export function createPreimage(challenge, miner) {
  const buf = new Uint8Array(PREIMAGE_BYTES);
  buf.set(hexToBytes(challenge, 32), 0);
  buf.set(hexToBytes(miner, 20), 32);
  return buf;
}

/** Write a uint256 into the preimage as 32 big-endian bytes, matching the EVM. */
export function writeNonce(buf, nonce) {
  let n = BigInt(nonce);
  if (n < 0n || n > MAX_U256) throw new Error("nonce out of uint256 range");
  for (let i = 31; i >= 0; i--) {
    buf[NONCE_OFFSET + i] = Number(n & 0xffn);
    n >>= 8n;
  }
  return buf;
}

/**
 * keccak256(abi.encodePacked(challenge, miner, nonce)) — the exact preimage
 * Nonce.sol hashes. Verified against on-chain vectors in test/parity.test.js.
 */
export function digest(challenge, miner, nonce) {
  return keccak_256(writeNonce(createPreimage(challenge, miner), nonce));
}

/** Difficulty score: type(uint256).max / uint256(digest), matching the contract. */
export function scoreOf(digestValue) {
  const d = typeof digestValue === "bigint" ? digestValue : bytesToBigInt(digestValue);
  if (d === 0n) return 0n;
  return MAX_U256 / d;
}

File: packages/cli/src/hash.rs
//! The proof-of-work hash.
//!
//! Must produce byte-identical output to `Nonce.sol`:
//!
//! ```solidity
//! keccak256(abi.encodePacked(bytes32 challenge, address miner, uint256 nonce))
//! ```
//!
//! That is 84 bytes: 32 challenge, 20 address, 32 big-endian nonce. One byte of
//! disagreement and every solution this miner finds is rejected on chain —
//! silently, because the failure looks like bad luck. `tests/parity.rs` checks
//! this against digest vectors the contract itself produced.

use primitive_types::U256;
use tiny_keccak::{Hasher, Keccak};

pub const PREIMAGE_LEN: usize = 32 + 20 + 32;
const NONCE_OFFSET: usize = 52;

/// Reusable preimage buffer for one (challenge, miner) pair.
///
/// The nonce is rewritten in place per attempt so the hot loop allocates
/// nothing; at a few hundred thousand hashes a second, an allocation per
/// attempt dominates the hash itself.
#[derive(Clone)]
pub struct Preimage {
    buf: [u8; PREIMAGE_LEN],
}

impl Preimage {
    pub fn new(challenge: [u8; 32], miner: [u8; 20]) -> Self {
        let mut buf = [0u8; PREIMAGE_LEN];
        buf[..32].copy_from_slice(&challenge);
        buf[32..52].copy_from_slice(&miner);
        Self { buf }
    }

    /// Write the nonce as 32 big-endian bytes, exactly as the EVM lays out a
    /// uint256. Always writes all 32 so a smaller value cannot inherit high
    /// bytes from the previous attempt.
    #[inline(always)]
    pub fn set_nonce(&mut self, nonce: U256) {
        nonce.to_big_endian(&mut self.buf[NONCE_OFFSET..]);
    }

    /// Fast path for the search loop: only the low 8 bytes change while the
    /// upper 24 stay fixed within a worker's slice.
    #[inline(always)]
    pub fn set_nonce_low(&mut self, low: u64) {
        self.buf[PREIMAGE_LEN - 8..].copy_from_slice(&low.to_be_bytes());
    }

    #[inline(always)]
    pub fn digest(&self) -> [u8; 32] {
        let mut out = [0u8; 32];
        let mut k = Keccak::v256();
        k.update(&self.buf);
        k.finalize(&mut out);
        out
    }

    pub fn bytes(&self) -> &[u8; PREIMAGE_LEN] {
        &self.buf
    }
}

/// `type(uint256).max / uint256(digest)` — the contract's difficulty score.
pub fn score_of(digest: &[u8; 32]) -> U256 {
    let d = U256::from_big_endian(digest);
    if d.is_zero() {
        U256::zero()
    } else {
        U256::MAX / d
    }
}

/// One-shot digest, for tests and verification rather than the hot loop.
pub fn digest(challenge: [u8; 32], miner: [u8; 20], nonce: U256) -> [u8; 32] {
    let mut p = Preimage::new(challenge, miner);
    p.set_nonce(nonce);
    p.digest()
}

pub fn keccak(input: &[u8]) -> [u8; 32] {
    let mut out = [0u8; 32];
    let mut k = Keccak::v256();
    k.update(input);
    k.finalize(&mut out);
    out
}
